src/topology/algebra/nonarchimedean/basic.lean

```diff
@@ -0,0 +1,171 @@
+/-
+Copyright (c) 2021 Ashwin Iyengar. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Kevin Buzzard, Johan Commelin, Ashwin Iyengar, Patrick Massot.
+-/
+import topology.algebra.ring
+import topology.algebra.open_subgroup
+import data.set.basic
+import group_theory.subgroup
+import algebra.ring.prod
+
+/-!
+# Nonarchimedean Topology
+
+In this file we set up the theory of nonarchimedean topological groups and rings.
+
+A nonarchimedean group is a topological group whose topology admits a basis of
+open neighborhoods of the identity element in the group consisting of open subgroups.
+A nonarchimedean ring is a topological ring whose underlying topological (additive)
+group is nonarchimedean.
+
+## Definitions
+
+- `nonarchimedean_add_group`: nonarchimedean additive group.
+- `nonarchimedean_group`: nonarchimedean multiplicative group.
+- `nonarchimedean_ring`: nonarchimedean ring.
+
+-/
+
+namespace set
+
+variables {α : Type*} {β : Type*}
+variables {s s₁ : set α} {t t₁ : set β}
+
+/-- If a product of s and t are subsets of P then the product of
+  two subsets of s and t (respectively) is a subset of P. -/
+lemma prod_subset_of_subsets {P : set (α × β)}:
+  s.prod t ⊆ P → s₁ ⊆ s → t₁ ⊆ t → s₁.prod t₁ ⊆ P :=
+begin
+  intros hP hs ht,
+  rw prod_subset_iff,
+  intros x hx y hy,
+  apply hP,
+  rw prod_mk_mem_set_prod_eq,
+  exact ⟨hs hx, ht hy⟩
+end
+
+end set
+
+/-- The cartesian product of two topological rings acquires a natural topology. -/
+instance (R S : Type*) [ring R] [ring S] [topological_space R] [topological_space S]
+  [topological_ring R] [topological_ring S] : topological_ring (R × S) :=
+{ continuous_neg := continuous_neg }
```

Comment:
This instance (along with one for semiring) should go in `topology/algebra/ring.lean`